File: src/list/MergeKSortedList.java
package list;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.PriorityQueue;

class Item implements Comparable<Item>{
	
	public Item(int data, int listID){
		this.data = data;
		this.listID = listID;
	}
	
	private int data;
	private int listID;
	
	public int getData(){return this.data;}
	public int getListID(){return this.listID;}
	
	public int compareTo(Item other) {
		return this.data - other.data;
	}
	
	public String toString(){
		return "(" + data + "," + listID + ")";
	}
	
}

public class MergeKSortedList {
	
	public static void main(String[] args) {
		ArrayList<Integer>[] lists = (ArrayList<Integer>[])new ArrayList[3];
		int [] a1 = {1,3,5};
		lists[0] = new ArrayList<Integer>(Arrays.asList(1,3,5));
		lists[1] = new ArrayList<Integer>(Arrays.asList(2,2,7));
		lists[2] = new ArrayList<Integer>(Arrays.asList(3,4,6));
		
		ArrayList<Integer> mList = mergeKSortedList(lists);
		//System.out.println(mList);
		
		String[] docs = new String[]{"1,2,3", "2,3", "2,5"};
		mr_reduce(docs);
	}
	
	public static void mr_reduce(String[] docs) {
		ArrayList<Integer>[] lists = (ArrayList<Integer>[])new ArrayList[docs.length];
		for (int i = 0; i < docs.length; i++) {
			lists[i] = new ArrayList<Integer>();
			String[] words = docs[i].split(",");
			for (int j = 0; j < words.length; j++) {
				lists[i].add(Integer.parseInt(words[j]));
			}
		}
		ArrayList<Integer> mList = mergeKSortedList(lists);
		LinkedHashSet<Integer> mList2 = new LinkedHashSet<Integer>(mList);
		for (Integer integer : mList2) {
			System.out.print(integer + " ");
		}
	}
	
	public static ArrayList<Integer> mergeKSortedList(ArrayList<Integer>[] lists){
		ArrayList<Integer> ret = new ArrayList<Integer>();
		PriorityQueue<Item> heap = new PriorityQueue<Item>();
		int[] ptrs = new int[lists.length];
		for (int i = 0; i < lists.length; i++) {
			//heap.offer(lists[i].get(0));
			heap.offer(new Item(lists[i].get(0), i));
			ptrs[i]++;
		}
		while(!heap.isEmpty()){
			Item t = heap.poll();
			ret.add(t.getData());
			int listID = t.getListID();
			if(ptrs[listID] < lists[listID].size()){
				heap.offer(new Item(lists[listID].get(ptrs[listID]), listID));
				ptrs[listID]++;
			}
		}
		return ret;
	}
}

File: src/list/merge_K_sorted_list.py
"""
Reducer
As you remember, the mappers previously produced key-value pairs of data - the key was the string we were searching for and the value was a list of matched documents.

We are now entering the Reducer phase of the MapReduce. Each reducer will receive the lists of matched documents produced by all mappers for a given string.
Our task is to simply merge the given lists of documents.
Given an array containing lists of sorted document IDs
Your task is to
write a function that prints to the standard output (stdout) the sorted document IDs obtained after merging the given lists
all IDs in the merged list must be unique
Note that your function will receive the following arguments:
docIds
which is an array of strings
each string contains a list of sorted document IDs separated by comma
Data constraints
all document IDs are integer numbers
each string in the docIds array will contain at most 100,000 IDs
the length of the docIds array will not exceed 1000 elements
Efficiency constraints
your function is expected to print the requested result and return in less than 2 seconds
expected complexity: O(N * logK), N = total number of IDs in all lists, K = total number of lists
Example
Input    Output
docIds: "1,2,3", "2,3", "2,5"
1 2 3 5
"""

#class HeapItem:
#    def __init__(self, word, docid):
#        self.word = word
#        self.docid = docid
#    def __eq__(self, other):
#        return self.word == other.word
#    def __lt__(self, other):
#        return self.word < other.word
        
docs = ["1,2,3", "2,3", "2,5"]

def mr_reduce(docs):
    res = []
    import Queue as Q
    q = Q.PriorityQueue() 
    docs = [map(int, doc.split(",")) for doc in docs]
    ptrs = [0 for i in xrange(len(docs))]
    for i, words in enumerate(docs):
        q.put((words[0],i))
        ptrs[i] += 1
    while not q.empty():
        e = q.get()
        print e
        res.append(e[0])
        docId = e[1]
        if( ptrs[docId] < len(docs[docId]) ):
            q.put((docs[docId][ptrs[docId]], docId))
            ptrs[docId] += 1
   
    for word in sorted(list(set(res))):
        print word
        
mr_reduce(docs)